waspc/src/Wasp/Generator/PackageJsonGenerator.hs

```diff
@@ -1,17 +1,71 @@
 module Wasp.Generator.PackageJsonGenerator
   ( resolveNpmDeps,
+    resolveDependencies,
+    DependencyConflictError (DependencyConflictError),
     npmDepsToPackageJsonEntry,
     npmDevDepsToPackageJsonEntry,
   )
 where
 
-import Data.Bifunctor (second)
-import Data.List (find, intercalate)
-import Data.Maybe (fromJust, isJust)
+import Data.List (intercalate)
+import qualified Data.Map as Map
+import qualified Data.Maybe as Maybe
 import qualified Wasp.AppSpec.App.Dependency as D
 
 type NpmDependenciesConflictError = String
 
+data DependencyConflictError = DependencyConflictError D.Dependency D.Dependency
+  deriving (Show, Eq)
+
+-- | Takes wasp npm dependencies and user npm dependencies and figures out how
+--   to combine them together, returning (Right) a list of npm dependencies to
+--   be used on behalf of wasp and then also a list of npm dependencies to be
+--   used on behalf of user. These lists might be the same as the initial ones,
+--   but might also be different.
+
+--   On error (Left), returns a list of conflicting wasp dep, user dep pairs
+
+--   This is used in resolveNpmDeps to separate the error message generation
+--   from its representation
+resolveDependencies ::
+  [D.Dependency] ->
+  [D.Dependency] ->
+  Either
+    [DependencyConflictError]
+    ([D.Dependency], [D.Dependency])
+resolveDependencies waspDeps userDeps =
+  if null conflicting
+    then Right (waspDeps, createDeps userDepsNotInWaspDeps)
+    else Left conflicting
+  where
+    waspDepsMap = createMap waspDeps
+    userDepsMap = createMap userDeps
+    overlappingDeps = waspDepsMap `Map.intersection` userDepsMap
+    userDepsNotInWaspDeps = userDepsMap `Map.difference` waspDepsMap
+    createMap :: [D.Dependency] -> Map.Map String D.Dependency
+    createMap deps = Map.fromList $ fmap toTuple deps
+      where
+        toTuple :: D.Dependency -> (String, D.Dependency)
+        toTuple dep = (D.name dep, dep)
+    createDeps :: Map.Map String D.Dependency -> [D.Dependency]
+    createDeps depsMap = Map.elems depsMap
+
+    -- get all items in overlappingDeps, for each check whether there's
+    -- a conflicting version in userDepsMap, and if so report a conflict
+    conflicting :: [DependencyConflictError]
+    conflicting = Maybe.mapMaybe createConflict (Map.assocs overlappingDeps)
+      where
+        createConflict :: (String, D.Dependency) -> Maybe DependencyConflictError
```

Comment:
Name confused me for a moment because I thought we are contructing/creating conflict, but I would say what we are actually doing is checking if conflict exists.

I would rename it to `findConflict` or `testForConflict` or `checkForConflict`. Or, to be really verbose: `checkIfWaspDepConflictsWithAnyUserDep`. Which might be the best, since this function is used at only one place and this makes it very easy to understand what it does.

Reply:
I don't like 'check' or 'test', as those imply bool to me, not returning a ConflictError. `find` is not really appropriate either, as we are really evaluating whether a potential conflict is indeed a conflict. I've gone for `makeConflictErrorIfMismatchedVersion`.

Reply:
Good point with `check` or `test`. I find that `find` actually works, in the sense of "we are scanning through user deps to find if there is any of them that conflicts the wasp dep". Also, `find` usually indicates return type is going to be Maybe or smth similar. But ok, no biggie, this is also fine.